### DISCUSSION TITLE
Getting the pose of UI inputs per frame

### DISCUSSION THREAD

**Discussion by austinbhale** [StereoKit/StereoKit#653] [Q&A]
hi nick, question for ya:
Is there somewhere at the end of a SK frame where I could find a list of all button ids (within their hierarchies) and their poses? Perhaps in the render queue. Or do you recommend to store the button poses outside of StereoKit? The main motivation is to simulate which button to press with the right hand like so:
![image](https://github.com/StereoKit/StereoKit/assets/32556981/1c3afc41-108f-431f-8b2d-edbe3b25db0a)

**Comment by maluoi** [comment]
Since StereoKit's UI is "Immediate Mode", the buttons don't really persist beyond the point where they are executed! You would indeed need to track the positions of them yourself.

If you need to simulate the button presses, there may be a more architectural solution?

```csharp
// Put active code in a separate function
if (UI.Button("Play")) PlayItem();

// That way it can be called based on other events!
if (simulationFrame == 12) PlayItem();

// Or even this, just make sure the UI element comes first :)
if (UI.Button("Play") || simulationFrame == 12) PlayItem();
```

If you're looking for a more complete testing simulation, you might record the hand poses, and play them back? I have a number of tools related to recording and replaying hands over in [DebugToolWindow.cs](https://github.com/StereoKit/StereoKit/blob/master/Examples/StereoKitTest/DebugToolWindow.cs). If your hand menu is head sensitive, you might need that too, which I'm pretty sure is doable as well?

**Reply by austinbhale** [reply]
![ezgif com-video-to-gif](https://github.com/StereoKit/StereoKit/assets/32556981/bb3804c1-8eb9-4199-a032-05aebed42e2d)

Ahh so SK is truly immediate mode, great! `UI.LayoutLast` ended up being enough to find the center positions 💯

(I didn't want to expose private methods and thought this type of automatic simulation would be more fun!)
